File: Toast/Sources/Toast/Helpers/Diagnostics.swift
import Foundation

struct DiagnosticSnapshot: Sendable {
    let appVersion: String
    let build: String
    let osVersion: String
    let arch: String
    let onboardingCompleted: Bool
    let projectsWatchedBucket: String
    let launchAtLogin: Bool
    let runInBackground: Bool
    let relaunchOnCrash: Bool
    let notificationsEnabled: Bool
    let showStatusText: Bool
    let analyticsEnabled: Bool
    let launcherRelaunch: Bool

    var properties: [String: Any] {
        [
            "app_version": appVersion,
            "build": build,
            "os_version": osVersion,
            "arch": arch,
            "onboarding_completed": onboardingCompleted,
            "projects_watched_bucket": projectsWatchedBucket,
            "launch_at_login": launchAtLogin,
            "run_in_background": runInBackground,
            "relaunch_on_crash": relaunchOnCrash,
            "notifications_enabled": notificationsEnabled,
            "show_status_text": showStatusText,
            "analytics_enabled": analyticsEnabled,
            "launcher_relaunch": launcherRelaunch,
        ]
    }
}

enum Diagnostics {
    static func projectsWatchedBucket(count: Int) -> String {
        switch count {
        case 0: "0"
        case 1: "1"
        case 2...5: "2-5"
        default: "6+"
        }
    }

    @MainActor
    static func snapshot(
        from store: DeploymentStore,
        launcherRelaunch: Bool = false
    ) -> DiagnosticSnapshot {
        DiagnosticSnapshot(
            appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown",
            build: Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "unknown",
            osVersion: ProcessInfo.processInfo.operatingSystemVersionString,
            arch: currentArchitecture,
            onboardingCompleted: store.hasCompletedOnboarding,
            projectsWatchedBucket: projectsWatchedBucket(count: store.watchedProjects.count),
            launchAtLogin: store.launchAtLoginEnabled,
            runInBackground: store.runInBackgroundEnabled,
            relaunchOnCrash: store.relaunchOnCrashEnabled,
            notificationsEnabled: store.notificationsEnabled,
            showStatusText: store.showStatusText,
            analyticsEnabled: AnalyticsService.shared.isEnabled,
            launcherRelaunch: launcherRelaunch
        )
    }

    private static var currentArchitecture: String {
        #if arch(arm64)
        "arm64"
        #elseif arch(x86_64)
        "x86_64"
        #else
        "unknown"
        #endif
    }
}

File: toast-win/src/Toast.Core/Diagnostics.cs
namespace Toast.Core;

public static class Diagnostics
{
    public static string ProjectsWatchedBucket(int count) => count switch
    {
        0 => "0",
        1 => "1",
        >= 2 and <= 5 => "2-5",
        _ => "6+",
    };

    public static string CurrentArchitecture()
    {
        return System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture switch
        {
            System.Runtime.InteropServices.Architecture.X64 => "x64",
            System.Runtime.InteropServices.Architecture.Arm64 => "arm64",
            System.Runtime.InteropServices.Architecture.X86 => "x86",
            _ => "unknown",
        };
    }

    public static Dictionary<string, object> Snapshot(
        bool onboardingCompleted,
        int projectsWatched,
        bool launchAtLogin,
        bool runInBackground,
        bool relaunchOnCrash,
        bool notificationsEnabled,
        bool showStatusText,
        bool analyticsEnabled,
        string appVersion,
        string build,
        bool launcherRelaunch = false)
    {
        return new Dictionary<string, object>
        {
            ["app_version"] = appVersion,
            ["build"] = build,
            ["os_version"] = Environment.OSVersion.VersionString,
            ["arch"] = CurrentArchitecture(),
            ["onboarding_completed"] = onboardingCompleted,
            ["projects_watched_bucket"] = ProjectsWatchedBucket(projectsWatched),
            ["launch_at_login"] = launchAtLogin,
            ["run_in_background"] = runInBackground,
            ["relaunch_on_crash"] = relaunchOnCrash,
            ["notifications_enabled"] = notificationsEnabled,
            ["show_status_text"] = showStatusText,
            ["analytics_enabled"] = analyticsEnabled,
            ["launcher_relaunch"] = launcherRelaunch,
        };
    }
}
